File: rust-2020/src/day02.rs
pub struct Field<'a> {
    lo: usize,
    hi: usize,
    chr: char,
    passwd: &'a str,
}

pub fn read(contents: &str) -> Vec<Field> {
    let mut data = Vec::<Field>::new();
    for line in contents.lines() {
        match *line
            .split([' ', '-'].as_ref())
            .collect::<Vec<_>>()
            .as_slice()
        {
            [lo, hi, chr, passwd] => {
                data.push(Field {
                    lo: lo.parse().unwrap(),
                    hi: hi.parse().unwrap(),
                    chr: chr.chars().next().unwrap(),
                    passwd,
                });
            }
            _ => unreachable!(),
        }
    }
    data
}

pub fn part1(data: &[Field]) -> usize {
    data.iter().fold(0, |acc, field| {
        let c = field.passwd.chars().filter(|&c| c == field.chr).count();
        if field.lo <= c && c <= field.hi {
            acc + 1
        } else {
            acc
        }
    })
}

pub fn part2(data: &[Field]) -> usize {
    data.iter().fold(0, |acc, field| {
        let loc: char = field.passwd.chars().nth(field.lo - 1).unwrap();
        let hic: char = field.passwd.chars().nth(field.hi - 1).unwrap();
        acc + ((loc == field.chr) ^ (hic == field.chr)) as usize
    })
}

pub fn run() -> (String, String) {
    let contents = include_str!(concat!(env!("CARGO_MANIFEST_DIR"), "/inputs/day02.input"));
    let data = read(&contents);
    let p1 = format!("{}", part1(&data));
    let p2 = format!("{}", part2(&data));
    (p1, p2)
}

File: 2020/day02.py
from collections import Counter

rules = []

with open("../inputs/2020/day02.input", "r") as f:
    for line in f.readlines():
        count, (char, _), passwd = line.strip().split(" ")
        lo, hi = map(int, count.split("-"))
        rules.append(((lo, hi), char, passwd))


def part1(rules):
    nvalid = 0
    for (lo, hi), char, passwd in rules:
        occurences = Counter(passwd).get(char, -1)
        if lo <= occurences <= hi:
            nvalid += 1
    return nvalid


def part2(rules):
    nvalid = 0
    for (lo, hi), char, passwd in rules:
        loc = passwd[lo-1]
        hic = passwd[hi-1]
        nvalid += (loc == char) ^ (hic == char)
    return nvalid


print(f"Part 1: {part1(rules)}")
print(f"Part 2: {part2(rules)}")
